packages/program/src/functions/multicall-scope.ts

```diff
@@ -40,4 +42,34 @@ export class MultiCallInvocationScope<TReturn = any> extends BaseInvocationScope
   addCalls(funcScopes: Array<FunctionInvocationScope>) {
     return super.addCalls(funcScopes);
   }
+
+  private validateHeapTypeReturnCalls() {
+    let heapOutputIndex: number | undefined;
+    let numberOfHeaps = 0;
+
+    this.calls.forEach((call, callIndex) => {
+      const { isOutputDataHeap } = call;
+
+      if (isOutputDataHeap) {
+        heapOutputIndex = callIndex;
+
+        if (++numberOfHeaps > 1) {
+          throw new FuelError(
+            ErrorCode.ONLY_ONE_HEAP_TYPE_CALL_ALLOWED,
+            'Only one call that returns a heap type is allowed on a multicall'
+          );
+        }
+      }
+    });
+
+    const hasHeapTypeReturn = typeof heapOutputIndex === 'number';
+    const isOnLastCall = heapOutputIndex === this.calls.length - 1;
```

Comment:
@camsjams, A _possible, unlikely, especially given the constructor requiring calls to be passed to it currently_ side-effect of making `heapOutputIndex`'s default value `-1` is that this `isOnLastCall` would evaluate to `true` and, if it was the only variable used below, the error would be thrown, which wouldn't make sense. I just wanted to throw it out there given that I've noticed it, for good measure.

Reply:
It won't work with just `isOnLastCall` because it will be undefined on multi-call without a call returning a heap type, and this will result `heapOutputIndex === this.calls.length - 1` being false.

The logic becomes a little messy because I've decided to mimic the Rust SDK implementation and throw errors for 2 different reasons. 

If I were to validate these 2 errors inside the loop, I would end up never actually throwing for more than 1 heap type return in the call since this would actually mean that one of these calls would not be at the last position in the multi-call. This means that one error would end up overlapping the other.

Based on the Rust SDK implementation I got under the impression that it is important to throw for these 2 different reasons.